app/_how-tos/autogenerate-serverless-mcp-tools.md

```diff
@@ -0,0 +1,218 @@
+---
+title: Autogenerate serverless MCP tools from a RESTful API
+content_type: how_to
+related_resources:
+  - text: AI Gateway
+    url: /ai-gateway/
+  - text: AI MCP
+    url: /plugins/ai-mcp/
+  - text: AI Proxy Advanced
+    url: /plugins/ai-proxy-advanced/
+
+description: Learn how to use the AI MCP plugin to generate serverless MCP from any RESTful API, including setting up a mock Node.js server for testing.
+products:
+  - gateway
+  - ai-gateway
+permalink: /mcp/autogenerate-serverless-mcp-tools/
+
+works_on:
+  - on-prem
+
+min_version:
+  gateway: '3.12'
+
+plugins:
+  - ai-mcp
+
+entities:
+  - service
+  - route
+  - plugin
+
+tags:
+  - ai
+  - openai
+  - mcp
+  - serverless
+
+tldr:
+  q: How do I automatically generate an MCP API from an existing REST API?
+  a: |
+    Use the AI MCP to map your REST API endpoints into MCP capabilities, allowing you to integrate them directly with AI Gateway.
+tools:
+  - deck
+
+prereqs:
+  inline:
+    - title: Cursor
+      content: |
+        This tutorial uses Cursor as an MCP client:
+        1. Go to the [Cursor downloads](https://cursor.com/downloads) page.
+        2. Download the installer for your operating system.
+        3. Install Cursor on your machine.
+        4. Launch Cursor and sign in to your account or create a new account.
+      icon_url: /assets/icons/cursor.svg
+  entities:
+    services:
+      - mcp-service
+    routes:
+      - mcp-route
+
+cleanup:
+  inline:
+    - title: Destroy the {{site.base_gateway}} container
+      include_content: cleanup/products/gateway
+      icon_url: /assets/icons/gateway.svg
+
+automated_tests: false
+---
+## Install mock API Server
+
+Before using the **AI MCP**, you’ll need an upstream HTTP API to expose. For this tutorial, we’ll use a simple mock API built with Express. This allows you to test the plugin without relying on an external service. This mock API simulates a small marketplace system with a fixed set of users and their associated orders. Each user has between two and five sample orders, which the API exposes through `/marketplace/users` and `/marketplace/{userId}/orders` endpoints.
+
+Running these commands will download the mock API script and install any required dependencies automatically:
+
+```sh
+curl -s -o api.js "https://gist.githubusercontent.com/subnetmarco/5ddb23876f9ce7165df17f9216f75cce/raw/a44a947d69e6f597465050cc595b6abf4db2fbea/api.js"
+npm install express
+node api.js
+```
+
+Validate the API is running:
+
+```sh
+curl -X GET http://localhost:3000
+```
+
+This request confirms that the mock server is up and responding. Later, the AI MCP will use this API’s OpenAPI schema to generate MCP tool definitions. You should see the following response from the server:
+
+```text
+{"name":"Sample Users API"}%
+```
+{:.no-copy-code}
+
+## Configure the AI MCP plugin
```

Comment:
In the current examples we are doing in-place protocol conversion, where the same route can handle both non-MCP request and MCP request.
I think we need an example when there're some ai-mcp plugin works in tool mode and one in server mode. This is useful for user wants to aggregate multiple different services in a new route.